python-lib/ontology_tagging/formatter_by_document.py

```diff
@@ -0,0 +1,442 @@
+
+    def _fill_tags(self, condition: bool, value: dict):
+        """Dump the dictionary 'value' if 'condition' is True"""
+        return json.dumps(value, ensure_ascii=False) if condition else np.nan
+
+    def write_df(
+        self,
+        input_df: pd.DataFrame,
+        text_column: AnyStr,
+        language_column: AnyStr = None,
+    ) -> pd.DataFrame():
+        """Write the output dataframe for format 'one row per document (array columns)' when there is no category in the ontology
+
+        Returns:
+            pd.DataFrame : a DataFrame with the following columns:
+            - all columns from the input dataframe input_df
+            - a column with the list of keywords
+            - a column with the list of tags
+            - a column with concatenated matched sentences
+
+        """
+        start = perf_counter()
+        self._generate_columns_names(input_df)
+        input_df.progress_apply(self._write_row, args=[language_column], axis=1)
+        logging.info(
+            f"Tagging {len(input_df)} documents : Done in {perf_counter() - start:.2f} seconds."
+        )
+        return self._merge_df_columns(input_df, text_column)
+
+    def _write_row(self, row: pd.Series, language_column: AnyStr = None) -> None:
+        """Called by write_df on each row
+        Append columns 'tag_sentences', 'tag_keywords' and 'tag_list' to the output dataframe self.output_df
+        Args:
+            row: pandas.Series from text_df
+            language_column: if not None, the matcher will be applied with the given language of the row
+        """
+        language = self._get_document_language(row, language_column)
+        document_to_match = self._get_document_to_match(row, language)
+        tags_in_document, keywords_in_document, matched_sentences = [], [], []
+        for idx, sentence in enumerate(document_to_match):
+            (
+                tags_in_document,
+                keywords_in_document,
+                matched_sentences,
+            ) = self._get_tags_in_row(
+                sentence=sentence,
+                original_sentence=row[self.text_column_tokenized][idx],
+                tags_in_document=tags_in_document,
+                keywords_in_document=keywords_in_document,
+                matched_sentences=matched_sentences,
+                language=language,
+            )
+
+        if tags_in_document != []:
+            line = {
+                self.tag_columns[0]: self._fill_tags(
+                    True, unique_list(tags_in_document)
+                ),
+                self.tag_columns[1]: self._fill_tags(
+                    True, unique_list(keywords_in_document)
+                ),
+                self.tag_columns[2]: "".join(unique_list(matched_sentences)),
+            }
+        else:
+            line = {column: np.nan for column in self.tag_columns}
+        self.output_df = self.output_df.append(line, ignore_index=True)
+
+    def _get_tags_in_row(
```

Comment:
This looks like a `_get_tags_in_sentence` more than `_get_tags_in_row`

Reply:
Right ! In PR #21